k2/python/k2/fsa_algo.py

```diff
@@ -131,6 +131,91 @@ def linear_fst(labels: Union[List[int], List[List[int]]],
     return fsa
 
 
+def linear_fst_with_self_loops(fsts: k2.Fsa):
+    '''Create a linear FST with epsilon self-loops by first removing epsilon
+    transitions from the input linear FST.
+
+
+    Note: The main difference to func:`linear_fsa_with_self_loops` is that
+          aux_labels and scores are also kept here.
+
+    Args:
+      fsas:
+        An FST or an FstVec. It MUST be a linear FST or a vector of linear FSTs.
+    Returns:
+      Return an FST or FstVec, where each FST contains epsilon self-loops but
+      contains no epsilon transitions for arcs that are not self-loops.
+    '''
+    assert hasattr(fsts, "aux_labels")
+    non_epsilon_positions = fsts.labels != 0
+    new_arc_flag = torch.ones(len(fsts.labels),
+                              dtype=torch.int32).to(fsts.device)
+
+    # The idea to generate dest_arc_row_ids:
+    # 1. if the input label of PREVIOUS src_arc is non zero,
+    #    there should be an new dest_arc in out_fst for CURRENT src_arc.
+    # 2. if the input label of PREVIOUS arc is zero,
+    #    CURRENT arc will share the dest_arc in out_fst of PREVIOUS src_arc.
+    # where src_arc is an arc in source fst,
+    # and dest_arc is corresponding arc in generated fst(out_fst).
+    #
+    #
+    # Following line works as:
+    # for i in range(1, len(fsts.labels)):
+    #     if(fsts.labels[i - 1] == 0):
+    #         new_arc_flag[i] = 0
+    #
+    # new_arc_flag only contains 0 or 1.
+    # 1 means:
+    # create a new dest_arc in out_fst for current src_arc.
+    # 0 means:
+    # current src_arc will share the dest_arc of previous src_arc.
+    new_arc_flag[1:] = non_epsilon_positions[0:-1]
+
+    dest_arc_row_ids = torch.cumsum(new_arc_flag, dim=0) - 1
```

Comment:
Could you explain why it can be used as `row_ids`?

Reply:
As in the comment:
    # new_arc_flag only contains 0 or 1.
    # 1 means:
    # create a new dest_arc in out_fst for current src_arc.
    # 0 means:
    # current src_arc will share the dest_arc of previous src_arc.
    
    If current src_arc has it's own dest_arc, then it should has its own row_id.
    If current src_arc share the dest_arc of previous src_arc, then it also share the row_id.
    
    So torch.cumsum(new_arc_flag) contains the same information of row_id, except the offset 1.